### PR TITLE
Add network creation and tests

### PR DESCRIPTION
**Pull request by dataclouder**
* Add Org Vdc network basic Creation function that returns a task (allows tight control on the creation)
* Add a customizable task monitoring function (same as WaitTaskCompletion, but with inspection hooks)
* Add monitoring functions for main library entities (monitor.go)
* Add function RemoveOrgVdcNetworkIfExists(networkName)
* Add tests for network creation: with Edge Gateway, with Direct connection to external network, isolated.
* Add parameter for external network in configuration file for tests
* Add clean-up code for networks in api_vcd_test.go
* Add variable to skip vApp creation in tests (used when running a single test that does not involve vApps)
* Fix various spelling mistakes in many code files
* Add test for external network retrieval

Signed-off-by: Giuseppe Maxia

Related issue: #49 

Preparatory work for Terraform provider issues [#96](https://github.com/terraform-providers/terraform-provider-vcd/issues/96) and [#95](https://github.com/terraform-providers/terraform-provider-vcd/issues/95)

### CODE REVIEW

File: govcd/orgvdcnetwork.go [hunk 1]
```diff
@@ -119,23 +157,19 @@ func (vdc *Vdc) CreateOrgVDCNetwork(networkConfig *types.OrgVDCNetwork) error {
...
+				return Task{}, fmt.Errorf("error decoding orgvdcnetwork response: %s", err)
 			}
-			task := NewTask(vdc.client)
-			for _, taskItem := range newstuff.OrgVDCNetwork.Tasks.Task {
-				task.Task = taskItem
-				err = task.WaitTaskCompletion()
-				if err != nil {
-					return fmt.Errorf("Error performing task: %#v", err)
-				}
+			for _, taskItem := range orgVDCNetwork.OrgVDCNetwork.Tasks.Task {
```
Comment (vbauzysvmware) [inline govcd/orgvdcnetwork.go]: I am not sure this is safe, cause there can be running previous task and we are not sure that first one is ours.

File: govcd/orgvdcnetwork_test.go [hunk 2]
```diff
@@ -31,3 +32,176 @@ func (vcd *TestVCD) Test_NetRefresh(check *C) {
...
+		vcd.org.Org.Name+"|"+vcd.vdc.Vdc.Name,
+		"Test_CreateOrgVdcNetworkIso")
+}
+
+// Tests the creation of a Org VDC network connected to an external network
+func (vcd *TestVCD) Test_CreateOrgVdcNetworkDirect(check *C) {
+	fmt.Printf("Running: %s\n", check.TestName())
+	networkName := TestCreateOrgVdcNetworkDirect
+
+	err := RemoveOrgVdcNetworkIfExists(vcd.vdc, networkName)
```
Comment (vbauzysvmware) [inline govcd/orgvdcnetwork_test.go]: I see some parts are good option for removing repeatable code.

File: govcd/orgvdcnetwork.go (hunk 1 above)
Comment (dataclouder) [reply govcd/orgvdcnetwork.go]: This one should be safe, as the network is the one we have just created, and there should not be any other task running. If there is, the whole operation would become meaningless, because we can only give back one task.
I will add a check to make sure there is only one active task, though.

File: govcd/orgvdcnetwork_test.go (hunk 2 above)
Comment (dataclouder) [reply govcd/orgvdcnetwork_test.go]: I have already squeezed the non-meaningful operations into functions. Since the file contains three different tests, I prefer leaving the details spelled out, as the tests can fail in different ways.

File: govcd/orgvdcnetwork.go [hunk 3]
```diff
@@ -90,23 +90,61 @@ func (orgVdcNet *OrgVDCNetwork) Delete() (Task, error) {
...
+	}
+	return nil
+}
+
+// Fine tuning network creation function.
+// Return an error (the result of the network creation) and a task (used to monitor
+// the network configuration)
+// This function can create any type of Org Vdc network. The exact type is determined by
+// the combination of properties given with the network configuration structure.
+func (vdc *Vdc) CreateOrgVDCNetWorkBasic(networkConfig *types.OrgVDCNetwork) (Task, error) {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/orgvdcnetwork.go]: I don't either. I was following other examples. I am open to suggestions to use a better naming scheme. In fact, the whole library would benefit from such naming re-organization.

**Comment by dataclouder** [comment]
Added a check for the task uniqueness after the network creation.

File: govcd/vdc.go [hunk 4]
```diff
@@ -125,6 +133,10 @@ func (vdc *Vdc) Refresh() error {
 
 func (vdc *Vdc) FindVDCNetwork(network string) (OrgVDCNetwork, error) {
 
+	err := vdc.Refresh()
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/vdc.go]: We have discussed this addition in Issue #105 (and in a recent meeting).
If we don't refresh, we get stale results, especially when we look for a network that was just created.
Either we do this refresh, or we should update the VDC object in memory after a creation. This kind of refresh is performed in other entities as well, and I assume it's done for the same reason.
In short, without a refresh we are using a VDC cache, without a way of knowing when the cache was invalidated. We can fix it, but it is not to be quick or pretty.

File: govcd/orgvdcnetwork.go [hunk 5]
```diff
@@ -60,7 +61,7 @@ func (orgVdcNet *OrgVDCNetwork) Refresh() error {
 func (orgVdcNet *OrgVDCNetwork) Delete() (Task, error) {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/orgvdcnetwork.go]: This is legacy code that had only cosmetic changes. I will add a comment anyway.

File: govcd/orgvdcnetwork.go [hunk 6]
```diff
@@ -90,23 +91,61 @@ func (orgVdcNet *OrgVDCNetwork) Delete() (Task, error) {
...
+		if err != nil {
+			return fmt.Errorf("error deleting network [task] %s", networkName)
+		}
+	}
+	return nil
+}
+
+// A wrapper call around CreateOrgVDCNetWorkBasic.
+// Creates a network and then uses the associated task to monitor its configuration
 func (vdc *Vdc) CreateOrgVDCNetwork(networkConfig *types.OrgVDCNetwork) error {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/orgvdcnetwork.go]: Thanks for the suggestion. I will apply this change

File: govcd/vdc.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hodgesrm) [reply govcd/vdc.go]: Any automatic solution to this needs to be validated by realistic performance tests.  We're adding what look like expensive network calls.  (I'm going to check into how we handle this issue for applications like the vCD UI, which presumably has this concern in spaces.)

[the remark this answers is not in the document]
Comment (lvirbalas) [reply govcd/vdc.go]: This is the way it is done currently throughout the library:
- In various vApp functions vdc.Refresh() is done;
- In various network functions vdc.Refresh() is done;
- In various firewall and VPN functions edgeGateway.Refresh() is done.

Given the above, let's do the following:
1.) Accept this vdc.Refresh() in FindVDCNetwork(...) function as well, as it makes it is both consistent with current model and, more importantly, easier for users to program with the lib.
2.) If we see a performance issue down the road, we fix it as such by adding a possibility to tune the behaviour of such functions by asking _not_ refresh explicitly.

Finally, there's not much value in making this refresh smart (by tracking the state of created resources), because new resources may be easily created _outside_ the library in:
a.) vCD GUI.
b.) Outside processes, like Terraform.
c.) Python CLI, etc.

[the remark this answers is not in the document]
Comment (lvirbalas) [reply govcd/vdc.go]: This is the way it is done currently throughout the library:
- In various vApp functions vdc.Refresh() is done;
- In various network functions vdc.Refresh() is done;
- In various firewall and VPN functions edgeGateway.Refresh() is done.

Given the above, let's do the following:
1.) Accept this vdc.Refresh() in FindVDCNetwork(...) function as well, as it makes it is both consistent with current model and, more importantly, easier for users to program with the lib.
2.) If we see a performance issue down the road, we fix it as such by adding a possibility to tune the behaviour of such functions by asking _not_ refresh explicitly.

Finally, there's not much value in making this refresh smart (by tracking the state of created resources), because new resources may be easily created _outside_ the library in:
a.) vCD GUI.
b.) Outside processes, like Terraform.
c.) Python CLI, etc.

[the remark this answers is not in the document]
Comment (lvirbalas) [reply govcd/vdc.go]: This is the way it is done currently throughout the library:
- In various vApp functions vdc.Refresh() is done;
- In various network functions vdc.Refresh() is done;
- In various firewall and VPN functions edgeGateway.Refresh() is done.

Given the above, let's do the following:
1.) Accept this vdc.Refresh() in FindVDCNetwork(...) function as well, as it makes it is both consistent with current model and, more importantly, easier for users to program with the lib.
2.) If we see a performance issue down the road, we fix it as such by adding a possibility to tune the behaviour of such functions by asking _not_ refresh explicitly.

Finally, there's not much value in making this refresh smart (by tracking the state of created resources), because new resources may be easily created _outside_ the library in:
a.) vCD GUI.
b.) Outside processes, like Terraform.
c.) Python CLI, etc.
